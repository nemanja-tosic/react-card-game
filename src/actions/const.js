export const PLAYER_PLAY_CARD = 'PLAYER_PLAY_CARD';
export const START_GAME = 'START_GAME';
export const NEW_GAME = 'NEW_GAME';
