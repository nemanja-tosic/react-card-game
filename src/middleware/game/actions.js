export const GAME_STARTING = 'GAME_STARTING';
export const GAME_UPDATE = 'GAME_UPDATE';

export const gameUpdate = (gameInfo) => ({
  type: GAME_UPDATE,
  payload: gameInfo,
});
