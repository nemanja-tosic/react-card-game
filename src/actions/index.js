import * as types from './const';

export const startGame = (playerName, numberOfOpponents) => ({
  type: types.START_GAME,
  payload: {
    playerName,
    numberOfOpponents,
  },
});

export const playCard = (cardIndex) => ({
  type: types.PLAYER_PLAY_CARD,
  payload: {
    cardIndex,
  },
});

export const newGame = () => ({
  type: types.NEW_GAME,
});
