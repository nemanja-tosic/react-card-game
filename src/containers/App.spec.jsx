// @flow
import React from 'react';
import configureStore from 'redux-mock-store';
import { shallow } from 'enzyme';

import * as actions from '../actions';
import AppContainer from './App';
import { TestGame } from '../test/support/test-game';
import GameComponent from '../components/Game';

describe('AppContainer', () => {
  let game;
  let store;

  describe('when the game has not started', () => {
    beforeEach(() => {
      game = TestGame.createNotStarted();
      store = configureStore()({
        game,
      });
    });

    it('should show the menu', () => {
      const appContainer = shallow(<AppContainer store={store}/>);

      expect(appContainer.getNode()).toMatchSnapshot();
    });

    it('should dispatch the NEW_GAME action when the onNewGame prop is called', () => {
      const playerName = 'player';
      const numberOfOpponents = 3;
      const appContainer = shallow(<AppContainer store={store}/>).dive();
      const gameComponent = appContainer.find('Menu');

      gameComponent.simulate('startGame', playerName, numberOfOpponents);

      expect(store.getActions()).toContainEqual(actions.startGame(playerName, numberOfOpponents));
    });
  });

  describe('when the game is loading', () => {
    beforeEach(() => {
      game = TestGame.createLoading();
      store = configureStore()({
        game,
      });
    });

    it('should show the loading screen', () => {
      const appContainer = shallow(<AppContainer store={store}/>);

      expect(appContainer.getNode()).toMatchSnapshot();
    });
  });

  describe('when the game is inProgress', () => {
    beforeEach(() => {
       game = TestGame.createInProgress();
       store = configureStore()({
        game,
      });
    });

    it('should show the game', () => {
      const appContainer = shallow(<AppContainer store={store}/>);

      expect(appContainer.getNode()).toMatchSnapshot();
    });

    it('should dispatch the card clicked action when the onCardClicked prop is called', () => {
      const appContainer = shallow(<AppContainer store={store}/>).dive();
      const gameComponent = appContainer.find('GameComponent');
      const cardIndex = 1;

      gameComponent.simulate('cardClicked', cardIndex);

      expect(store.getActions()).toContainEqual(actions.playCard(cardIndex));
    });
  });

  describe('when the game is over', () => {
    beforeEach(() => {
      game = TestGame.createOver();
      store = configureStore()({
        game,
      });
    });

    it('should show the scoreboard', () => {
      const appContainer = shallow(<AppContainer store={store}/>);

      expect(appContainer.getNode()).toMatchSnapshot();
    });

    it('should dispatch the NEW_GAME action when the onNewGame prop is called', () => {
      const appContainer = shallow(<AppContainer store={store}/>).dive();
      const gameComponent = appContainer.find('Scores');

      gameComponent.simulate('newGame');

      expect(store.getActions()).toContainEqual(actions.newGame());
    });
  });
});
