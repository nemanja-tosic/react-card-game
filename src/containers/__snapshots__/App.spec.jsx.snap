// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppContainer when the game has not started should show the menu 1`] = `
<App
  game={
    Game {
      "players": Array [
        Player {
          "cards": Array [
            Card {
              "code": "0C",
              "imageUrl": "https://deckofcardsapi.com/static/img/0C.png",
            },
            Card {
              "code": "2C",
              "imageUrl": "https://deckofcardsapi.com/static/img/2C.png",
            },
            Card {
              "code": "3C",
              "imageUrl": "https://deckofcardsapi.com/static/img/3C.png",
            },
            Card {
              "code": "4C",
              "imageUrl": "https://deckofcardsapi.com/static/img/4C.png",
            },
            Card {
              "code": "5C",
              "imageUrl": "https://deckofcardsapi.com/static/img/5C.png",
            },
          ],
          "id": 1,
          "isActive": true,
          "isUserControlled": true,
          "name": "Player",
          "score": 0,
          "takenCards": Array [],
        },
      ],
      "state": "notStarted",
      "talon": Array [
        Array [
          Card {
            "code": "JS",
            "imageUrl": "https://deckofcardsapi.com/static/img/JS.png",
          },
        ],
      ],
      "winners": Array [],
    }
  }
  handleCardClicked={[Function]}
  handleNewGame={[Function]}
  handleStartGame={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "notify": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": null,
    }
  }
/>
`;

exports[`AppContainer when the game is inProgress should show the game 1`] = `
<App
  game={
    Game {
      "players": Array [
        Player {
          "cards": Array [
            Card {
              "code": "0C",
              "imageUrl": "https://deckofcardsapi.com/static/img/0C.png",
            },
            Card {
              "code": "2C",
              "imageUrl": "https://deckofcardsapi.com/static/img/2C.png",
            },
            Card {
              "code": "3C",
              "imageUrl": "https://deckofcardsapi.com/static/img/3C.png",
            },
            Card {
              "code": "4C",
              "imageUrl": "https://deckofcardsapi.com/static/img/4C.png",
            },
            Card {
              "code": "5C",
              "imageUrl": "https://deckofcardsapi.com/static/img/5C.png",
            },
          ],
          "id": 1,
          "isActive": true,
          "isUserControlled": true,
          "name": "Player",
          "score": 0,
          "takenCards": Array [],
        },
      ],
      "state": "inProgress",
      "talon": Array [
        Array [
          Card {
            "code": "JS",
            "imageUrl": "https://deckofcardsapi.com/static/img/JS.png",
          },
        ],
      ],
      "winners": Array [],
    }
  }
  handleCardClicked={[Function]}
  handleNewGame={[Function]}
  handleStartGame={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "notify": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": null,
    }
  }
/>
`;

exports[`AppContainer when the game is loading should show the loading screen 1`] = `
<App
  game={
    Game {
      "players": Array [
        Player {
          "cards": Array [
            Card {
              "code": "0C",
              "imageUrl": "https://deckofcardsapi.com/static/img/0C.png",
            },
            Card {
              "code": "2C",
              "imageUrl": "https://deckofcardsapi.com/static/img/2C.png",
            },
            Card {
              "code": "3C",
              "imageUrl": "https://deckofcardsapi.com/static/img/3C.png",
            },
            Card {
              "code": "4C",
              "imageUrl": "https://deckofcardsapi.com/static/img/4C.png",
            },
            Card {
              "code": "5C",
              "imageUrl": "https://deckofcardsapi.com/static/img/5C.png",
            },
          ],
          "id": 1,
          "isActive": true,
          "isUserControlled": true,
          "name": "Player",
          "score": 0,
          "takenCards": Array [],
        },
      ],
      "state": "loading",
      "talon": Array [
        Array [
          Card {
            "code": "JS",
            "imageUrl": "https://deckofcardsapi.com/static/img/JS.png",
          },
        ],
      ],
      "winners": Array [],
    }
  }
  handleCardClicked={[Function]}
  handleNewGame={[Function]}
  handleStartGame={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "notify": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": null,
    }
  }
/>
`;

exports[`AppContainer when the game is over should show the scoreboard 1`] = `
<App
  game={
    Game {
      "players": Array [
        Player {
          "cards": Array [
            Card {
              "code": "0C",
              "imageUrl": "https://deckofcardsapi.com/static/img/0C.png",
            },
            Card {
              "code": "2C",
              "imageUrl": "https://deckofcardsapi.com/static/img/2C.png",
            },
            Card {
              "code": "3C",
              "imageUrl": "https://deckofcardsapi.com/static/img/3C.png",
            },
            Card {
              "code": "4C",
              "imageUrl": "https://deckofcardsapi.com/static/img/4C.png",
            },
            Card {
              "code": "5C",
              "imageUrl": "https://deckofcardsapi.com/static/img/5C.png",
            },
          ],
          "id": 1,
          "isActive": true,
          "isUserControlled": true,
          "name": "Player",
          "score": 0,
          "takenCards": Array [],
        },
      ],
      "state": "over",
      "talon": Array [
        Array [
          Card {
            "code": "JS",
            "imageUrl": "https://deckofcardsapi.com/static/img/JS.png",
          },
        ],
      ],
      "winners": Array [],
    }
  }
  handleCardClicked={[Function]}
  handleNewGame={[Function]}
  handleStartGame={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "notify": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": null,
    }
  }
/>
`;
